### DISCUSSION TITLE
Producing custom array sharding

### DISCUSSION THREAD

**Discussion by luyug** [jax-ml/jax#23703] [Q&A]
I am trying to define some array shardings in jax. Typically I will use `NamedSharding` but I am working on a sharding scheme that I presume not captured by NamedSharding.

As an example, say I have a (8 x 8) array and (2, 2) device mesh. I want to shard the array such that if it is reshaped into (2 x 2 x 2 x 8), the 2-nd and the 4-th dimension will be sharded on the 1st and the 2nd dimensions of the mesh, except the sharding should work with the original 2d array.

I wonder what is an elegant / jax-ish way to do it?

**Comment by yashk2810** [comment]
Just to clarify, you want to shard your original array in a block-cyclic way to begin with so that when you reshape it to (2, 2, 2, 8) you don't get any comms right?

If yes, then there is no way to do that right now but we have thought about this problem before since block-cyclic shardings end up looking like this. We would need a new representation in PartitionSpec that encodes block-cyclic shardings kind of like your example. We have had some ideas on how to make this work with NamedSharding. We think that it's doable though.

You can also write your own sharding by subclassing `jax.sharding.Sharding` and then implementing `_to_sdy_sharding` to lower it to the underlying IR (which will be pretty similar to how NamedSharding._to_sdy_sharding` looks like). Thought in this case, AFAIK, shardy doesn't support strided sharding yet.

P.S. Sorry for the late reply. Credit where it's due, I found this issue linked from ezyang's blog: https://blog.ezyang.com/2025/12/learning-to-love-mesh-oriented-sharding/

**Reply by yashk2810** [reply]
To distill your question further, it can be:

Let's say you have a mesh: `{'x': 2}`, If you have an input `a: f32[2, 2@x]` and you want to reshape to `(4,)` what should the output sharding look like so that there is *no communication*? As it stands right now, there will be comms since the output will be strided. But if we have strided sharding support, then one example of how the output spec can look like is `P((<stride_offset/placeholder>, 'x'))` to avoid comms.

Let me know if I got this interpretation wrong :)

**Reply by yashk2810** [reply]
FWIW, you can implement a sort of interleaved reshape inside a shard_map to get an interleaved array with no comms. Here's an example

```
In [34]: mesh = jax.make_mesh((2,), ('x',), axis_types=(jax.sharding.AxisType.Explicit,))

In [35]: jax.set_mesh(mesh)
Out[35]: <jax._src.sharding_impls.set_mesh at 0x7fe5e4346c20>

In [36]: arr = jax.device_put(jnp.arange(4).reshape(2, 2), P(None, 'x'))

In [39]: arr
Out[39]:
Array([[0, 1],
            [2, 3]], dtype=int32)

In [37]: @jax.shard_map(in_specs=P(None, 'x'), out_specs=P('x'))
    ...: def f(x):
    ...:     return x.reshape((2,))
    ...:

In [38]: f(arr)
Out[38]: Array([0, 2, 1, 3], dtype=int32)

In [40]: f(arr).sharding
Out[40]: NamedSharding(mesh=Mesh('x': 2, axis_types=(Explicit,)), spec=PartitionSpec('x',), memory_kind=device)
```

This keeps the output sharding as `P('x')` but as you can see the output array is different from the input array (if you flatten it) i.e. it's interleaved. Probably not what you want but just thought of putting it out there.

**Reply by yashk2810** [reply]
@pschuh pointed out that you can also do `arr.T.reshape((4,))` to get the same interleaved result which makes sense. But the shmap technique can be applied more generally in other cases I think.
